File: govenus/comm/mailbox_utils.go
package comm

import "fmt"

type sniffer struct {
	action func(Message)
}

func NewSniffer() Mailbox {
	return &sniffer{action: func(m Message) {
		fmt.Println(m)
	}}
}

func NewCustomSniffer(action func(Message)) Mailbox {
	return &sniffer{action: action}
}

func (sniffer *sniffer) Notify(message Message, _ Client) {
	sniffer.action(message)
}

File: rsvenus/src/comm/mailbox_utils.rs
use super::{Mailbox, Message, Client};

pub enum Sniffer {
  Default,
  Custom(Box<dyn Fn(Message) + Send + Sync + 'static>)
}

impl Sniffer {
  pub fn custom<T : Fn(Message) + Send + Sync + 'static>(action: T) -> Self {
    Self::Custom(Box::new(action))
  }
}

impl Default for Sniffer {
  fn default() -> Self {
    Self::Default
  }
}

impl Mailbox for Sniffer {
  fn notify(&mut self, message: Message, _: Client) {
    match self {
      Self::Default => println!("{:?}", message),
      Self::Custom(action) => action(message),
    }
  }
}